connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java

```diff
@@ -793,6 +796,38 @@ protected void assignConnectors(List<WorkerLoad> workerAssignment, Collection<St
         }
     }
 
+    static class BalancedIterator<E> implements Iterator<E> {
+
+        Map<String, Iterator<E>> grouped;
+        List<String> keys;
+        int k;
+
+        public BalancedIterator(Collection<E> collection, Function<E, String> allocationGrouper) {
+            this.k = 0;
+            this.grouped = collection.stream()
+                .collect(Collectors.groupingBy(item -> allocationGrouper.apply(item)))
+                .entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().iterator()));
```

Comment:
Nits (can use `allocationGrouper` directly instead of `item -> allocationGrouper.apply(item)`, and can avoid creating multiple streams by using the two-argument variant of `Collectors::groupingBy`):
```suggestion
            this.grouped = collection.stream().collect(Collectors.groupingBy(
                    allocationGrouper,
                    Collectors.collectingAndThen(
                            Collectors.toList(),
                            List::iterator
                    )
            ));
```